presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueryFramework.java

```diff
@@ -43,31 +45,69 @@
 import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
 import static java.lang.String.format;
 import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.fail;
 
 public abstract class AbstractTestQueryFramework
 {
-    protected final H2QueryRunner h2QueryRunner;
-    protected final QueryRunner queryRunner;
-    private final SqlParser sqlParser;
-
+    private QueryRunnerSupplier queryRunnerSupplier;
+    private final Closer closer = Closer.create();
+    protected QueryRunner queryRunner;
+    private H2QueryRunner h2QueryRunner;
+    private SqlParser sqlParser;
+    private boolean initialized;
+
+    @Deprecated
     protected AbstractTestQueryFramework(QueryRunner queryRunner)
     {
-        this.queryRunner = queryRunner;
-        h2QueryRunner = new H2QueryRunner();
-        sqlParser = new SqlParser();
+        this(() -> queryRunner, true);
     }
 
-    @AfterClass(alwaysRun = true)
-    public void close()
+    protected AbstractTestQueryFramework(QueryRunnerSupplier supplier)
+    {
+        this(supplier, false);
+    }
+
+    private AbstractTestQueryFramework(QueryRunnerSupplier supplier, boolean initialize)
+    {
+        this.queryRunnerSupplier = requireNonNull(supplier, "queryRunnerSupplier is null");
+        // TODO refactor all the tests and remove this hack
+        if (initialize) {
+            init();
+        }
+    }
+
+    @BeforeClass
+    public void init()
     {
+        if (initialized) {
+            return;
+        }
         try {
-            h2QueryRunner.close();
+            queryRunner = closer.register(queryRunnerSupplier.get());
         }
-        finally {
-            queryRunner.close();
+        catch (InterruptedException e) {
```

Comment:
Before the migration it was being called from the test constructor. I've no idea what TestNG does during the initialization. So, it is always better to be safe =) But now yes, i can remove this entirely, and just add `throws Exception` declaration to the `init()` method.